clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -527,4 +536,87 @@ public void onComplete(ClientResponse response) {
         boolean shouldBlock();
     }
 
+
+    /*
+     * A threadsafe helper class to hold requests per node that has not been sent yet
+     */
+    private final static class UnsentRequests {
+        private final Map<Node, List<ClientRequest>> unsent;
+
+        public UnsentRequests() {
+            unsent = new HashMap<>();
+        }
+
+        public synchronized void put(Node node, List<ClientRequest> requests) {
+            List<ClientRequest> nodeUnsent = unsent.get(node);
+            if (nodeUnsent == null) {
+                nodeUnsent = new ArrayList<>();
+                unsent.put(node, nodeUnsent);
+            }
+            nodeUnsent.addAll(requests);
+        }
+
+        public synchronized void put(Node node, ClientRequest request) {
+            List<ClientRequest> nodeUnsent = unsent.get(node);
+            if (nodeUnsent == null) {
+                nodeUnsent = new ArrayList<>();
+                unsent.put(node, nodeUnsent);
+            }
+            nodeUnsent.add(request);
+        }
+
+        public synchronized int getRequestCount(Node node) {
```

Comment:
We don't usually use the `get` prefix in Kafka. This can probably just be `requestCount` right? Same for the other `getRequestCount` and `getNodes`.